### PR TITLE
enh_diar joint model

### PR DESCRIPTION
**Pull request by YushiUeda**
This PR is a reimplementation of https://github.com/espnet/espnet/pull/4234, which intends to incorporate `enh_diar` into the `enh_s2t` framework (https://github.com/espnet/espnet/pull/4226).

TODOs:
- [x] add README
  - [ ] add results
  - [ ] upload pretrained models
- [x] add test codes

### CODE REVIEW

File: espnet2/bin/enh_scoring_flexible_numspk.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
```
Comment (simpleoier) [inline espnet2/bin/enh_scoring_flexible_numspk.py:1]: This file has a lot overlap with espnet2/bin/enh_scoring.py. Can you add an option to inform flexible numspk in enh_scoring.py? Then just change the corresponding part.

File: espnet2/diar/loss/wrappers/flexible_spk_pit_solver.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+from itertools import permutations
```
Comment (simpleoier) [inline espnet2/diar/loss/wrappers/flexible_spk_pit_solver.py:1]: This file has a lot of overlap as espnet2/enh/loss/wrapper/pit_solver.py. Probably you can combine those two? The difference is only that num_spk determined by inf or ref. Maybe set an option to choose?

**Review by simpleoier — CHANGES_REQUESTED** [review]

File: espnet2/enh/espnet_enh_s2t_model.py [hunk 3]
```diff
@@ -224,15 +273,47 @@ def encode(
...
         """
-        speech_pre, feature_mix, feature_pre, others = self.enh_model.forward_enhance(
-            speech, speech_lengths
-        )
+        (
+            speech_pre,
+            feature_mix,
+            feature_pre,
+            others,
+            _,
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/espnet_enh_s2t_model.py:281]: Yes, I agree.

File: egs2/TEMPLATE/enh_diar1/enh_diar.sh [hunk 4]
```diff
@@ -0,0 +1,897 @@
...
+max_wav_duration=    # Maximum duration in second
+
+# diar model related
+diar_tag=    # Suffix to the result dir for diar enh model training.
+diar_config= # Config for diar model training.
+diar_args=   # Arguments for diar model training, e.g., "--max_epoch 10".
+             # Note that it will overwrite args in diar config.
+feats_normalize=utterance_mvn # Normalizaton layer type.
+spk_num=2    # Number of speakers in the input audio
+noise_type_num=1
```
[the remark this answers is not in the document]
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/enh_diar.sh:59]: Added the explanation to the help message.

File: egs2/TEMPLATE/enh_diar1/enh_diar.sh [hunk 5]
```diff
@@ -0,0 +1,897 @@
...
+dereverb_ref_num=1
+
+# Training data related
+use_dereverb_ref=false
```
[the remark this answers is not in the document]
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/enh_diar.sh:63]: ditto

File: espnet2/enh/espnet_enh_s2t_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (YushiUeda) [reply espnet2/enh/espnet_enh_s2t_model.py]: Included bottleneck outputs in `others`

File: egs2/librimix/enh_diar1/local/data.sh [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+log "data preparation started"
+
+# Github LibriMix : https://github.com/s3prl/LibriMix.git
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ] ; then
```
[the remark this answers is not in the document]
Comment (YushiUeda) [reply egs2/librimix/enh_diar1/local/data.sh:39]: I prefer to keep the current code since they are slightly different. `enh_diar` has extra files for the enhancement task (`spk{1,2,3}.scp` and `noise1.scp`)

File: espnet2/bin/enh_scoring_flexible_numspk.py (hunk 1 above)
Comment (YushiUeda) [reply espnet2/bin/enh_scoring_flexible_numspk.py]: Added an option to enh_scoring.py

File: espnet2/diar/loss/wrappers/flexible_spk_pit_solver.py (hunk 2 above)
Comment (YushiUeda) [reply espnet2/diar/loss/wrappers/flexible_spk_pit_solver.py]: Added an option to pit_solver.py

**Comment by YushiUeda** [comment]
@simpleoier thanks for your review and helpful suggestions! I believe I've addressed all of your points.

File: egs2/TEMPLATE/enh_diar1/enh_diar.sh [hunk 7]
```diff
@@ -0,0 +1,905 @@
...
+                    '{ if ($2 > min_length ) print $0; }' \
+                    >"${data_feats}/${dset}/utt2num_samples"
+            for spk in ${_spk_list} "wav"; do
+                <"${data_feats}/org/${dset}/${spk}.scp" \
+                    utils/filter_scp.pl "${data_feats}/${dset}/utt2num_samples"  \
+                    >"${data_feats}/${dset}/${spk}.scp"
+            done
+
+            # fix_data_dir.sh leaves only utts which exist in all files
+            #utils/fix_data_dir.sh --utt_extra_files "${_scp_list}" "${data_feats}/${dset}"
```
Comment (simpleoier) [inline egs2/TEMPLATE/enh_diar1/enh_diar.sh:322]: Can you remove the commented lines if they are not used?

File: egs2/TEMPLATE/enh_diar1/enh_diar.sh [hunk 8]
```diff
@@ -0,0 +1,905 @@
...
+            utils/fix_data_dir.sh "${data_feats}/${dset}"
+            #sort spk{i}.scp, etc.
```
Comment (simpleoier) [inline egs2/TEMPLATE/enh_diar1/enh_diar.sh:324]: ditto.

File: egs2/librimix/enh_diar1/conf/tuning/train_diar_enh_convtasnet_adapt.yaml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+label_aggregator: label_aggregator
+label_aggregator_conf:
+    win_length: 256
+    hop_length: 64  # This value should be set to:
+                    # enh_encoder stride * subsampling factor (corresponding to diar_encoder input_layer)
+                    # subsampling factors: 1 (linear), 2 (conv2d2), 4 (conv2d), 6 (conv2d6), 8 (conv2d8)
+diar_frontend: null
+frontend: null # this is only required when "diar_front_end: null"
+# uncomment the following lines to use concatenated log-mel filterbank and bottleneck features
+# diar_frontend: default
```
Comment (simpleoier) [inline egs2/librimix/enh_diar1/conf/tuning/train_diar_enh_convtasnet_adapt.yaml:70]: Can you clean the config?

File: egs2/librimix/enh_diar1/local/prepare_kaldifiles.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
+import os
```
Comment (simpleoier) [inline egs2/librimix/enh_diar1/local/prepare_kaldifiles.py:1]: Does this file cover all the purposes of `egs2/librimix/diar1/local/prepare_diarization.py`? If so, maybe two files can be merged?

File: egs2/librimix/enh_diar1/local/run_adapt.sh [hunk 11]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
```
Comment (simpleoier) [inline egs2/librimix/enh_diar1/local/run_adapt.sh:1]: Where is this file used?

**Review by simpleoier — COMMENTED** [review]
It looks in a good shape. I left some small comments.

File: egs2/TEMPLATE/enh_diar1/enh_diar.sh (hunk 7 above)
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/enh_diar.sh]: removed them

File: egs2/TEMPLATE/enh_diar1/enh_diar.sh (hunk 8 above)
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/enh_diar.sh]: ditto

File: egs2/librimix/enh_diar1/conf/tuning/train_diar_enh_convtasnet_adapt.yaml (hunk 9 above)
Comment (YushiUeda) [reply egs2/librimix/enh_diar1/conf/tuning/train_diar_enh_convtasnet_adapt.yaml:70]: Should I split it into 2 files? I could otherwise write some instructions in README.

File: egs2/librimix/enh_diar1/local/prepare_kaldifiles.py (hunk 10 above)
Comment (YushiUeda) [reply egs2/librimix/enh_diar1/local/prepare_kaldifiles.py:1]: There will be some extra files (spk & noise scp files) if we use this file for diarization. If that is OK, we can merge them.

File: egs2/librimix/enh_diar1/local/run_adapt.sh (hunk 11 above)
Comment (YushiUeda) [reply egs2/librimix/enh_diar1/local/run_adapt.sh:1]: This is used for fine-tuning the model on the 2&3-speaker mixture dataset. I will add some instructions in README later.

Comment (simpleoier) [reply egs2/librimix/enh_diar1/local/run_adapt.sh:1]: Got it!

File: egs2/librimix/enh_diar1/local/prepare_kaldifiles.py (hunk 10 above)
Comment (simpleoier) [reply egs2/librimix/enh_diar1/local/prepare_kaldifiles.py:1]: I don't know, maybe some flags to decide whether to generate the extra files? Or in `data.sh`, remove these files if not necessary.

File: egs2/librimix/enh_diar1/conf/tuning/train_diar_enh_convtasnet_adapt.yaml (hunk 9 above)
Comment (simpleoier) [reply egs2/librimix/enh_diar1/conf/tuning/train_diar_enh_convtasnet_adapt.yaml:70]: Maybe spliting it into 2 files is better.

**Comment by YushiUeda** [comment]
Hi @simpleoier, all the checks have passed, is it possible to merge this PR at this point?
I was planning to add results and pretrained models in this PR, but the experiments seem to take a while because of the very long queue time.

File: espnet2/bin/enh_scoring.py [hunk 12]
```diff
@@ -29,6 +29,7 @@ def scoring(
     ref_scp: List[str],
```
Comment (sw005320) [inline espnet2/bin/enh_scoring.py:29]: It’s good to make a unit test and cover this function.

File: espnet2/enh/espnet_model.py [hunk 13]
```diff
@@ -169,14 +172,38 @@ def forward_enhance(
         additional: Optional[Dict] = None,
     ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
         feature_mix, flens = self.encoder(speech_mix, speech_lengths)
-        feature_pre, flens, others = self.separator(feature_mix, flens, additional)
+        if self.mask_module is None:
```
Comment (sw005320) [inline espnet2/enh/espnet_model.py:175]: I’m actually not sure what is the mask module.
Can you explain it to me?

File: espnet2/diar/separator/tcn_separator_nomask.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
+from distutils.version import LooseVersion
```
Comment (sw005320) [inline espnet2/diar/separator/tcn_separator_nomask.py:1]: Please explain the difference between tcn_separator.py and this one.

File: espnet2/diar/layers/tcn_nomask.py [hunk 15]
```diff
@@ -0,0 +1,264 @@
+# Implementation of the TCN proposed in
```
Comment (sw005320) [inline espnet2/diar/layers/tcn_nomask.py:1]: Please explain the difference between tcn.py and this one

**Review by sw005320 — COMMENTED** [review]
- There are four training data configs. It is better to explain what they are (in README.md?)
- It’s better to explain what is `run.sh` and what is `run_adapt.sh` (in README.md?)
- Add README.md and upload models later
- It’s good to add an integration test

File: espnet2/enh/espnet_model.py (hunk 13 above)
Comment (YushiUeda) [reply espnet2/enh/espnet_model.py:175]: Let me clarify what `mask_module` and `tcn_separator` are here.
`mask_module` corresponds to the final 1x1 conv block and nonlinear activation function in `tcn_separator` for estimating the masks. On the other hand, `tcn_separator_nomask` is `tcn_separator` without the mask estimation part.
I split `tcn_separator` into `tcn_separator_nomask` and `mask_module` so that:
- the intermediate bottleneck features (output of `tcn_separator_nomask`) can be fed to the diarization branch (the original `tcn_separator` does not output such features)
- masks corresponding to the given number of speakers are estimated (the number of speakers is fixed in the `tcn_separator`)

Comment (sw005320) [reply espnet2/enh/espnet_model.py:175]: Got it!
Thanks for the clarification.
I think you may already do that, but it would be great to leave such comments in the source code.

Comment (YushiUeda) [reply espnet2/enh/espnet_model.py:175]: Thanks! Will do.

**Comment by YushiUeda** [comment]
@sw005320 thank you for your comments. I believe I've covered them.

File: egs2/TEMPLATE/enh_diar1/scripts/utils/create_README_file.py [hunk 16]
```diff
@@ -0,0 +1 @@
+egs2/TEMPLATE/asr1/scripts/utils/create_README_file.py
```
[the remark this answers is not in the document]
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/scripts/utils/create_README_file.py]: Thanks. It should have been `../../../asr1/scripts/utils/create_README_file.py`

File: egs2/TEMPLATE/enh_diar1/scripts/utils/get_model_names.py [hunk 17]
```diff
@@ -0,0 +1 @@
+egs2/TEMPLATE/asr1/scripts/utils/get_model_names.py
```
[the remark this answers is not in the document]
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/scripts/utils/get_model_names.py]: ditto

File: egs2/TEMPLATE/enh_diar1/scripts/utils/score_der.sh [hunk 18]
```diff
@@ -0,0 +1 @@
+../../../diar1/scripts/utils/score_der.sh
```
[the remark this answers is not in the document]
Comment (YushiUeda) [reply egs2/TEMPLATE/enh_diar1/scripts/utils/score_der.sh:1]: This one was actually OK.

File: espnet2/bin/enh_scoring.py (hunk 12 above)
Comment (YushiUeda) [reply espnet2/bin/enh_scoring.py:29]: Added the test

**Comment by kamo-naoyuki** [comment]
I changed ci tests to check the import order by [isort](https://github.com/PyCQA/isort) #4372, so please also apply isort after resolving the conflicts. Sorry for bothering.

**Comment by sw005320** [comment]
@YushiUeda, after you fix the above conflicts, I’ll merge this PR.

**Comment by sw005320** [comment]
Thanks, @YushiUeda!
